client/my-sites/plugins/plugins-list/index.jsx

```diff
@@ -0,0 +1,402 @@
+			sites: PropTypes.array,
+			slug: PropTypes.string,
+			name: PropTypes.string,
+			wpcom: PropTypes.bool,
+			wporg: PropTypes.bool,
+		} ) ).isRequired,
+		header: PropTypes.string.isRequired,
+		isWpCom: PropTypes.bool.isRequired,
+		sites: PropTypes.object.isRequired,
+		selectedSite: PropTypes.oneOfType( [ PropTypes.bool, PropTypes.object ] ).isRequired,
+		pluginUpdateCount: PropTypes.number,
+		isPlaceholder: PropTypes.bool.isRequired,
+	},
+
+	getInitialState() {
+		return {
+			disconnectJetpackDialog: false,
+			bulkManagement: false,
+			selectedPlugins: {}
+		};
+	},
+
+	componentDidMount() {
+		if ( this.props.isWpCom !== true ) {
+			PluginsLog.on( 'change', this.showDisconnectDialog );
+		}
+	},
+
+	componentWillUnmount() {
+		if ( this.props.isWpCom !== true ) {
+			PluginsLog.removeListener( 'change', this.showDisconnectDialog );
+		}
+	},
+
+	isSelected( plugin ) {
+		return !! this.state.selectedPlugins[ plugin.slug ];
+	},
+
+	togglePlugin( plugin ) {
+		const { slug } = plugin;
+		const { selectedPlugins } = this.state;
+		const oldValue = selectedPlugins[ slug ];
+		this.setState( { selectedPlugins: Object.assign( {}, selectedPlugins, { [ slug ]: ! oldValue } ) } );
+		analytics.ga.recordEvent( 'Plugins', 'Clicked to ' + this.isSelected( plugin ) ? 'Deselect' : 'Select' + 'Single Plugin', 'Plugin Name', slug );
+	},
+
+	setBulkSelectionState( plugins, selectionState ) {
+		let slugsToBeUpdated = {};
+		plugins.forEach( plugin => slugsToBeUpdated[ plugin.slug] = this.hasNoSitesThatCanManage( plugin ) ? false : selectionState );
+
+		this.setState( { selectedPlugins: Object.assign( {}, this.state.selectedPlugins, slugsToBeUpdated ) } );
+	},
+
+	getPluginBySlug( slug ) {
+		for ( let i = 0, l = this.props.plugins.lenght; i < l; i++ ) {
+			if ( this.props.plugins[ i ].slug === slug ) {
+				return this.props.plugins[ i ];
+			}
+		}
+		return false;
+	},
+
+	filterSelection: {
+		active( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.active );
+			}
+			return false;
+		},
+		inactive( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && ! site.plugin.active );
+			}
+			return false;
+		},
+		updates( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.update && site.canUpdateFiles );
+			}
+			return false;
+		},
+		selected( plugin ) {
+			return this.isSelected( plugin );
+		}
+	},
+
+	getPluginUpdateCount() {
+		if ( this.props.isWpCom ) {
+			return 0;
+		}
+		return this.props.pluginUpdateCount;
+	},
+
+	hasNoSitesThatCanManage( plugin ) {
+		return ! plugin.wpcom &&
+			! plugin.sites.some( site => site.modules && site.modules.indexOf( 'manage' ) !== -1 );
+	},
+
+	getSelected() {
+		if ( ! this.props.plugins ) {
+			return [];
+		}
+		return this.props.plugins.filter( this.filterSelection.selected.bind( this ) );
```

Comment:
A note on how `filter` works: if `this.props.plugins` is an empty list, `filter()` will return that empty list here too. Not sure if `this.props.plugins` can be null or undefined, but if it's always a list we can skip the error checking conditional and just return the filter.